apiextensions-apiserver: add columns to CRD spec

sttts opened the pull request:
Follow-up of https://github.com/kubernetes/kubernetes/pull/60269.

```release-note
Add spec.additionalPrinterColumns to CRDs to define server side printing columns.
```

sttts committed:
apiextensions-apiserver: add columns to CRD spec

sttts committed:
Update generated files

sttts commented:
/cc @kubernetes/sig-api-machinery-api-reviews

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -34,6 +34,16 @@ type CustomResourceDefinitionSpec struct {
 	// Subresources describes the subresources for CustomResources
 	Subresources *CustomResourceSubresources
+
+	// Columns
```
this does not even belong under validation.

lcfang commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -425,7 +425,7 @@ func (r *crdHandler) getOrCreateServingInfoFor(crd *apiextensions.CustomResource
 	}
 
 	// TODO: identify how to pass printer specification from the CRD
```
can we delete this "TODO"?

sttts replied:
Yes, we should.

jennybuckley commented:
/cc @yliaog

sttts commented:
/assign @soltysh @juanvallejo can you comment on the API?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -34,6 +34,16 @@ type CustomResourceDefinitionSpec struct {
+	// Header is the header to be printed on-top of the column.
+	Header string
+	// Path is a simple JSON path, i.e. with array notation.
+	Path string
```
Do we want to support `Type string` with OpenAPI types? Currently every cell is returned as string.

soltysh replied:
I'd start with strings, for now. We can always improve if needed.

smarterclayton replied:
We can't actually return new cells of different types in the future unless we define it now because that would not be backwards compatible for a client that assumes strings.  We need to up front force people to acknowledge non-strings if we ever want to support them.

soltysh reviewed (changes requested):
This lgtm, but trying this out I can't get this to work, so probably there's some more wiring needed. I can look into it tomorrow.

soltysh commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/tableconvertor/tableconvertor.go:
```diff
@@ -33,50 +31,33 @@ import (
-		if err := path.Parse(parts[1]); err != nil {
-			return c, fmt.Errorf("unrecognized column definition in 'x-kubernetes-print-columns': %v", spec)
+
+	for _, col := range crdColumns {
```
I've played a bit with the PR and the only comment that I have is this. If a CRD author specifies columns we should *overwrite* the default ones. This will provide a better UX for CRD authors, otherwise they might end up with unwanted columns. If you look closely [here](https://github.com/kubernetes/kubernetes/blob/1d905bbdfcb07c2660f7c70a1e9d4111285436f1/pkg/printers/internalversion/printers.go) how we specify the columns for the built-in types we always specify full set, there's no defaulting. I'd imagine this being the same for CRDs.

sttts replied:
Should we default the field to the default columns?

smarterclayton replied:
I don't think we should allow them to overwrite name.  I'm ok with forcing them to specify createdAt, and that would actually be better because that's an example of a date field we should handle correctly.

Defaulting the field may make sense, not 100% sure.

sttts replied:
if we enforce name and createAt, I suggest to rename this field to `AdditionalColumns` and don't default it.

sttts replied:
Or even `AdditionalPrinterColumns`.

soltysh replied:
Let's default and allow them to change it later on. I don't agree with @smarterclayton here about not having the ability to drop `name` field. The author should have that option, and defaulting is exactly the right place for that.

soltysh reviewed (dismissed):
The overall state of the PR lgtm, although I'd like to hear comments from @kubernetes/sig-api-machinery-api-reviews

smarterclayton commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -34,6 +34,16 @@ type CustomResourceDefinitionSpec struct {
+	Columns []Column
+}
+
+type Column struct {
```
CustomResourceColumnDefinition

smarterclayton replied:
To match `TableColumnDefinition` in meta/v1beta1

sttts replied:
done

smarterclayton commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -34,6 +34,16 @@ type CustomResourceDefinitionSpec struct {
+	// Header is the header to be printed on-top of the column.
+	Header string
+	// Path is a simple JSON path, i.e. with array notation.
+	Path string
```
Call this `JSONPath`

sttts replied:
done

smarterclayton commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/tableconvertor/tableconvertor.go:
```diff
@@ -107,7 +88,57 @@ func (c *convertor) ConvertToTable(ctx genericapirequest.Context, obj runtime.Ob
+						cells = append(cells, cell)
+						continue
+					}
+					// TODO: types.go for Cells mentiones "simple maps", but resttest.go forbid them. Clarify!
```
Simple maps was intended for things like labels, but we haven't gotten there yet.  I think the important outcome is that clients *must* hide/handle types they don't recognize, but we have to realize that not all clients will do that.

sttts replied:
I have restricted the types for CRDs via validation to the primitive types for now: string, bool, integer, number.

dims commented:
some of the failures in verify look legit?

sttts commented:
@dims haven't checked. this needs another round anyway to incorporate @smarterclayton's comments.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/table_test.go:
```diff
@@ -0,0 +1,183 @@
+		if got, expected := tbl.Rows[0].Cells[3], int64(10); got != expected {
+			t.Errorf("expected cell[3] to equal %#v, got %#v", expected, got)
+		}
+		if got, expected := tbl.Rows[0].Cells[4], "bar"; got != expected {
```
compare this behaviour: the cells are not rejected (we don't have a way in the table types to return an error, skipping and object feels wrong), but the wrong type is just passed along.

Do we want that? If not, what do we want instead?

smarterclayton replied:
I would say we should pass `<error>` as a string for now.  We can always change later (output is not stable).

smarterclayton replied:
Actually, let's just pass nil for that particular cell.  Nil cells should always be allowed.

sttts replied:
done

sttts commented:
@smarterclayton @soltysh completed the PR and addressed your comments.

This question is open: what should we do if a field in a CR does not match the specified type of the column? Now, I just pass it through, making the client responsible for error checking or graceful handling of type errors.

sttts commented:
/cc @nikhita

sttts commented:
/retest

lavalamp commented:
/assign @mbohlool

soltysh commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -34,6 +34,31 @@ type CustomResourceDefinitionSpec struct {
 	Subresources *CustomResourceSubresources
+
+	// AdditionalPrinterColumns are additional columns shown e.g. in kubectl next to the name and creation time.
+	AdditionalPrinterColumns []CustomResourceColumnDefinition
```
It looks like I'm overruled here. So I'm guessing the always present field will be `name` and this will be defaulted to `createdAt`?

sttts replied:
fixed

soltysh replied:
Nit: s/created-at/createdAt

soltysh commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -31,6 +32,12 @@ import (
+var (
+	openAPIv3DataTypes                    = sets.NewString("integer", "number", "string", "boolean", "object", "array")
+	openAPIv3PrimitiveDataTypes           = sets.NewString("integer", "number", "string", "boolean")
+	oustomResourceColumnDefinitionFormats = sets.NewString("int32", "int64", "float", "double", "byte", "date", "date-time", "password")
```
Typo: **c**ustomResourceColumnDefinitionFormats

sttts replied:
fixed

soltysh reviewed (dismissed):
A few nits and the missing defaulting and I'm ok with this PR.

sttts commented:
@soltysh ptal at the second commit, addressing your nits (and adding some more complexity working around that the age is special elsewhere in the code).

soltysh reviewed (approved):
/lgtm
Squash the new commit with the old one and unless @smarterclayton objects this is good for merge

sttts commented:
/assign @liggitt 

To approve the swagger.json change.

sttts commented:
/retest

smarterclayton commented:
/approve
